src/passes/Print.cpp

```diff
@@ -169,13 +182,16 @@ struct SExprType {
   SExprType(Type type) : type(type){};
 };
 
-static std::ostream& operator<<(std::ostream& o, const SExprType& sType) {
+static std::ostream& printSExprType(std::ostream& o,
+                                    const SExprType& sType,
+                                    Module* wasm = nullptr) {
   Type type = sType.type;
   if (type.isTuple()) {
     o << '(';
     auto sep = "";
     for (const auto& t : type) {
-      o << sep << SExprType(t);
+      o << sep;
+      printSExprType(o, t, wasm);
```

Comment:
I do kind of like having these custom printers as subexpressions rather than as their own statements, even when they're not implemented via wrapper types. Being uniform about this would be a bigger separate change, though.

```suggestion
      o << sep << printSExprType(o, t, wasm);
```

Reply:
Interesting. I think that could work, but we'd need to remove the `o` parameter and have it return the result, I think (to avoid nondeterminism in what is called first if there are multiple calls on a single line).

I also agree that maybe a single big Printer class might help here. I'm not sure though.

For both, I think I'd prefer to leave them to future investigation/refactoring.

Separately, I think this PR should be complete now, after I figured out a nondeterminism bug.